Npgsql/NpgsqlLargeObjectStream.cs

```diff
@@ -0,0 +1,306 @@
+    /// Note that the OpenRead/OpenReadWrite method as well as all operations performed on this stream must be wrapped inside a database transaction.
+    /// </summary>
+    public class NpgsqlLargeObjectStream : Stream
+    {
+        NpgsqlLargeObjectManager _manager;
+        uint _oid;
+        int _fd;
+        long _pos;
+        bool _writeable;
+        bool _disposed;
+
+        private NpgsqlLargeObjectStream() { }
+
+        internal NpgsqlLargeObjectStream(NpgsqlLargeObjectManager manager, uint oid, int fd, bool writeable)
+        {
+            _manager = manager;
+            _oid = oid;
+            _fd = fd;
+            _pos = 0;
+            _writeable = writeable;
+        }
+
+        void CheckDisposed()
+        {
+            if (_disposed)
+                throw new InvalidOperationException("Object disposed");
+        }
+
+        /// <summary>
+        /// Reads up to <i>count</i> bytes from the large object into the buffer.
+        /// A return value of 0 indicates end of file.
+        /// </summary>
+        /// <param name="buffer">The buffer where read data should be stored.</param>
+        /// <param name="offset">The offset in the buffer where the first byte should be stored.</param>
+        /// <param name="count">The maximum number of bytes that should be read.</param>
+        /// <returns>How many bytes actually read, or 0 if end of file was already reached.</returns>
+        [GenerateAsync]
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset < 0)
+                throw new ArgumentOutOfRangeException("offset");
+            if (count < 0)
+                throw new ArgumentOutOfRangeException("count");
+            if (buffer.Length - offset < count)
+                throw new ArgumentException("Invalid offset or count for this buffer");
+            Contract.EndContractBlock();
+
+            CheckDisposed();
+
+            count = Math.Min(count, _manager.MaxTransferBlockSize);
+
+            var bytesRead = _manager.ExecuteFunctionGetBytes("loread", buffer, offset, count, _fd, count);
+            _pos += bytesRead;
+            return bytesRead;
+        }
+
+        /// <summary>
+        /// Reads <i>count</i> bytes from the large object. The only case when fewer bytes are read is when end of stream is reached.
+        /// </summary>
+        /// <param name="buffer">The buffer where read data should be stored.</param>
+        /// <param name="offset">The offset in the buffer where the first byte should be stored.</param>
+        /// <param name="count">The maximum number of bytes that should be read.</param>
+        /// <returns>How many bytes actually read, or 0 if end of file was already reached.</returns>
+        [GenerateAsync]
+        public int ReadAll(byte[] buffer, int offset, int count)
```

Comment:
Actually after looking at this a bit more... Can you explain why MaxTransferBlockSize is needed? Is it some sort of performance optimization?

If you want to keep MaxTransferBlockSize, then IMHO the ReadAll() function is what should be exposed as Read() to the user. In other words, the MaxTransferBlockSize is a totally internal performance detail in the Stream (even if the user can set it), and shouldn't affect what the user receives when calling Read(). So if the user calls read with count > MaxTransferBlockSize, and the Read() function receives a whole chunk (i.e. with size MaxTransferBlockSize), it should continue on to the next chunk (like in ReadAll). However, if at any point Read() internally doesn't receive the amount of data from the socket that it requested (return value of ExecuteFunctionGetBytes is smaller than its count parameter), then it should return immediately. This is the standard Stream API behavior.

But again, I think it may be better to drop MaxTransferBlockSize altogether...